DBZ-1238 Allow more options for snapshot SPI

addisonj opened the pull request:
https://issues.jboss.org/projects/DBZ/issues/DBZ-1238

This commit adds some more options to the Postgres Snapshottter SPI
interface, specifically, the ability to create the query for the start
of a snapshot, as well as the ability to generate a statement to lock
the tables.

Additionally, some rework is done to allow for the usage of an exported
snapshot in conjuction with the SPI taking metadata about a newly
created slot. This also changes how slots are initialized to allow for
the exported snapshot to be referenced, which requires adding a more
granular setup of a ReplicationConnection

Together, these options allow for a lot more flexibility, such as not
taking locks when snapshotting large databases as well as potentially to
use exported snapshots.

A future version should perhaps make some options possible without
implementing the SPI (such as using the exported snapshot by default)
but this commit is just trying to get all the groundwork in place to
allow for the Snapshotter SPI to have this functionality.

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -994,14 +1009,15 @@ protected static ConfigDef configDef() {
         Field.group(config, "Connector", TOPIC_SELECTION_STRATEGY, CommonConnectorConfig.POLL_INTERVAL_MS, CommonConnectorConfig.MAX_BATCH_SIZE, CommonConnectorConfig.MAX_QUEUE_SIZE,
                     CommonConnectorConfig.SNAPSHOT_DELAY_MS,
-                    SNAPSHOT_MODE, SNAPSHOT_LOCK_TIMEOUT_MS, TIME_PRECISION_MODE, DECIMAL_HANDLING_MODE, HSTORE_HANDLING_MODE, SCHEMA_REFRESH_MODE, ROWS_FETCH_SIZE);
+                    SNAPSHOT_MODE, SNAPSHOT_LOCK_TIMEOUT_MS, TIME_PRECISION_MODE, DECIMAL_HANDLING_MODE, HSTORE_HANDLING_MODE,
```
I moved these around as I think they were grouped improperly before... not sure if that breaks any assumptions in KafkaConnect, but it is better organized

gunnarmorling commented:
Hi @addisonj, wouldn't this be a potential use case for the new snapshotting class SPI?

addisonj commented:
It could potentially... but at the moment, the SPI just controls to start the snapshot and/or streaming consumer and generates the queries.

If you wanted to extend the idea to also controlling more aspects of the snapshot, that might make sense, but just adding an option to take locks or not feels a bit strange, as it is pretty specific for what is easily an option.

Perhaps the SPI could also have a method for taking a connection and a list of tables to open and close the transaction? That feels pretty flexible, but IDK of how different the use cases would be...

gunnarmorling commented:
> but just adding an option to take locks or not feels a bit strange, as it is pretty specific for what is easily an option.

I'm wondering whether this just shouldn't be a new *value* for the existing "snapshot mode* option. Or is it rather an orthogonal concern to the existing snapshotting modes and could be combined with any of them?

gunnarmorling commented:
Any thoughts on ^^^, @addisonj?

addisonj commented:
Hey Gunnar, yes I have seen both of these, just got diverted onto something
different for a bit, hope to get back to this before end of week!

On Tue, Apr 23, 2019 at 2:15 AM Gunnar Morling <notifications@github.com>
wrote:

> Any thoughts on ^^^, @addisonj <https://github.com/addisonj>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/debezium/debezium/pull/852#issuecomment-485691906>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE3LA3TO62EP4OVFES57CDPR3AR7ANCNFSM4HGB5VBQ>
> .
>

gunnarmorling commented:
Hey, grooming the backlog a bit, I came across two related issues:

* https://issues.jboss.org/browse/DBZ-1252: this suggests to establish "lock mode" as an orthogonal concern to snapshotting mode -- something we already have for SQL Server. This seems pretty reasonable, "no lock" would then just be one specific value of that new option
* https://issues.jboss.org/browse/DBZ-1035: this is about leveraging "exported snapshots"; in the light of the issue above, this could be another locking strategy chooseable from that option for lock modes

Curious what you think about it?

addisonj commented:
just getting back to this, thanks for pointing that out @gunnarmorling, will try and look deeper and see if there isn't a way to accommodate that with the SPI

addisonj commented:
@gunnarmorling I think this should be ready, I ended up having to change a few tests that I think just were passing because of luck and not very specific assertions before, otherwise, I think this introduces everything needed to make those new options *possible* via the SPI, but not integrated directly as options (which could be done relatively quickly as a follow on).

Let me know what you think, I don't have a ton of time in the near future to keep working on this, but let me know if there is anything you were thinking of a lot differently.

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -85,64 +92,51 @@ private PostgresReplicationConnection(Configuration config,
-            throw e;
-        }
-        catch (Throwable t) {
+            ensureSlotNotActive(getSlotInfo());
```
This is probably the most tricky change in the whole commit, rather than always connecting right away, we now delay connecting which allows us to have more control over the lifecycle of when a snapshot is created, specifically, this needed to happen so that when you a export a snapshot, no additional queries are done on the slot until the snapshot is done, otherwise, that invalidates the exported snapshot.

gunnarmorling replied:
Gotcha; thanks.

gunnarmorling commented:
@addisonj, getting back to this now after two weeks of conference travel. Need to wrap my head around what's happening again.

gunnarmorling commented:
@addisonj, reading the notion of "exported snapshots" a few times. Does this mean that this now addresses DBZ-1035, too? Taking a step back, could you perhaps describe on a somewhat higher level what has changed here? Thanks a lot, that'd help to undestand the gist of this.

addisonj commented:
hi @gunnarmorling, no worries, hope it was a nice time :) 

So yes, this is intended to exported snapshots (DBZ-1035) possible. It is not turning them on by default or exposing options to use the, but all of the pipes should be connected to do so, but I considered this change big enough already!

As far as an overview, I added a description to the commit message (copied here) but since I resused the old PR, the description didn't get updated (I just edited it):
```
This commit adds some more options to the Postgres Snapshottter SPI
interface, specifically, the ability to create the query for the start
of a snapshot, as well as the ability to generate a statement to lock
the tables.

Additionally, some rework is done to allow for the usage of an exported
snapshot in conjuction with the SPI taking metadata about a newly
created slot. This also changes how slots are initialized to allow for
the exported snapshot to be referenced, which requires adding a more
granular setup of a ReplicationConnection

Together, these options allow for a lot more flexibility, such as not
taking locks when snapshotting large databases as well as potentially to
use exported snapshots.

A future version should perhaps make some options possible without
implementing the SPI (such as using the exported snapshot by default)
but this commit is just trying to get all the groundwork in place to
allow for the Snapshotter SPI to have this functionality.
```

In other words it:
- adds the new methods to the SPI interface for generating the transaction and lock queries
- changes how a replicationConnection is initiated, which allows for us to use an exportedSnapshot (you can't make any other commands/queries on the connection after the slot is created, or it invalidates the snapshot)

gunnarmorling commented:
> hi @gunnarmorling, no worries, hope it was a nice time :)

Thanks, stressful yet insightful nevertheless :)

Digesting this change as we speak, definitely a lot is happening here...

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java:
```diff
@@ -100,14 +100,15 @@ protected SlotState getCurrentSlotInfo() throws SQLException {
     }
 
-    protected ReplicationConnection createReplicationConnection() throws SQLException {
+    protected ReplicationConnection createReplicationConnection(Boolean exportSnapshot) throws SQLException {
```
The parameter type of `exportSnapshotOnCreate()` is a primitive `boolean`, so it should be the same here for the sake of consistency and avoiding potential NPEs.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java:
```diff
@@ -172,37 +190,33 @@ private void takeSnapshot(BlockingConsumer<ChangeEvent> consumer) {
+                schema.refresh(connection, false);
+            }
+            else {
+                logger.warn("Step 2: skipping locking each table, this may result in inconsistent schema!");
```
Is this actually true when using exported snapshots? It seems that message might make things sound worse than they are in this case?

gunnarmorling commented on debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/OutboxEventRouterIT.java:
```diff
@@ -121,6 +121,12 @@ public void shouldConsumeRecordsFromInsert() throws Exception {
         assertThat(routedEvent).isNotNull();
         assertThat(routedEvent.topic()).isEqualTo("outbox.event.user");
+
+        Struct valueStruct = requireStruct(routedEvent.value(), "test payload");
```
That's interesting; can't see the link to the actual change here, so I'm curious about it (not to say it's bad)?

addisonj replied:
I don't have a link... but after a bit of debugging, I realized the problem, specifically, before, the test was actually not using specific enough assertions and multiple messages would work, notice that this test asserted on anything more than the topic. This occurred  because the slot could have old state due to DROP_SLOT_ON_STOP being false. 

This change, mostly just due to adding new tests, made that different enough that it did fail some of the later tests. 

By making this test drop the slot and then have more specific assertions, we made it a lot more stable

gunnarmorling replied:
Ok, gotcha. Makes sense. Thanks!

gunnarmorling commented:
@addisonj, I actually went ahead and pushed a commit to address some of the things I've commented on.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -198,12 +205,82 @@ public ReplicationStream startStreaming(Long offset) throws SQLException, Interr
+
     private ReplicationStream createReplicationStream(final LogSequenceNumber lsn) throws SQLException, InterruptedException {
         PGReplicationStream s;
+        initConnection();
```
I think this might be a problem; see the two sub-sequent calls to `initReplicationSlot()` later in this method. They are in try/catch blocks do work-around issues with different versions of wal2json which might reject the slot options we pass. We wouldn't get to these catch blocks any longer due to this call here. This is all a bit messy unfortunately.

That said, what's the need for `initConnection()` to begin with? That's not quite clear to me yet?

addisonj replied:
This actually is currently wrong (about to push up a commit to fix it) but not quite for the reasons listed.

The exceptions below happen when **starting** the slot, not when creating the slot. When you create a slot, you simply specify the plugin, but you don't pass any options. The `initConnection` makes sure the slot exists and isn't actively being used, and also fetches the defaultStartingPosition (which is what is wrong here, we need to call initConnection earlier).

So all his code should be safe to do here

gunnarmorling commented:
@addisonj, ok, pushed another commit with some small fixes. On a high level, these are the things I see:

* needs rebase to use the new common snapshot.fetch.size option
* initialization of the slot without try/catch might be an issue for wal2json (see inline comment)
* the interactions in `PostgresReplicationConnection` are getting more and more complex; there's multiple initXy(), startStream(), createStream() methods and it's tough to grok what's going on. I think this needs a major overhaul eventually (not necessarily as part of this PR, but it seems this adds on the complexity that already existed before)

gunnarmorling commented:
Hey @addisonj, quick heads-up: we're planning to do the 0.10.0.Alpha1 release tomorrow. Beta1 will follow up on that rather quickly, so whenever this here is ready we can include it into a release soon thereafter. There's also related work going for unifying more code of the connectors, which might impact the design here (e.g. see the `SnapshotChangeEventSource` contract already used for Oracle and SQL Server). But let's cross that bridge when we get there (migrating the PG connector to those framework classes), after all the SPI here is marked as "incubating".

addisonj commented:
Hi Gunnar,

Okay, tomorrow I was planning on addressing all the things mentioned, but
if it looks like it will be a bit, then I may not rush as much, but will
still try and get everything addressed ASAP.

Thanks,

Addison

On Mon, May 27, 2019 at 10:23 AM Gunnar Morling <notifications@github.com>
wrote:

> Hey @addisonj <https://github.com/addisonj>, quick heads-up: we're
> planning to do the 0.10.0.Alpha1 release tomorrow. Beta1 will follow up on
> that rather quickly, so whenever this here is ready we can include it into
> a release soon thereafter. There's also related work going for unifying
> more code of the connectors, which might impact the design here (e.g. see
> the SnapshotChangeEventSource contract already used for Oracle and SQL
> Server). But let's cross that bridge when we get there (migrating the PG
> connector to those framework classes), after all the SPI here is marked as
> "incubating".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/debezium/debezium/pull/852?email_source=notifications&email_token=AAE3LAYUK7N42PXU2LOUEWDPXQDHBA5CNFSM4HGB5VB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWKFGLA#issuecomment-496259884>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE3LAZBCX3ZKOHFF3Q7G5LPXQDHBANCNFSM4HGB5VBQ>
> .
>

gunnarmorling commented:
Yes, no need to rush it, the next release (0.10 Beta1 is planned for next week).

One quick other question: I noticed "xmin" was added to the "source" block of Postgres events produced by Debezium (DBZ-1082). Is this information actually relevant for consumers, or should this rather have gone only into the (connector-internal) offset structure?

addisonj commented:
@gunnarmorling yes, it was intended to be exposed, as that is the useful bit we are using to write our own snapshotter that allows us to ensure we don't miss any changes in the event of a slot fail-over

gunnarmorling commented:
>  as that is the useful bit we are using to write our own snapshotter that allows us to ensure we don't miss any changes in the event of a slot fail-over

I see, thanks. Could you expand a bit perhaps on the usage? Do you retrieve the value manually from already produced messages in Kafka? I'm wondering whether it perhaps could even be a short guest post on our blog describing this approach a bit? WDYT?

addisonj commented:
@gunnarmorling I pushed an updated fix to this and this realized I made a goof! I forgot you had pushed some changes, so I think I force pushed over them. I am not sure how to get them back as I didn't pull the branch before force pushing. Do you still have those locally? 

That said I think I addressed most your comments (other than the ones I brought back...), but just general things:
- I added a few comments to make it more clear, but the initConnection was in the wrong place (needed to happen earlier on), but it is safe to do with the wal2json stuff as creating and starting the slot in distinct
- As for a refactor, I agree it could be cleaned up some, but it did feel better suited to a follow on, which I can perhaps help with, but for now, I think it mostly represents the real complexity, given that creating a slot, init-int a slot, and starting a slot are all distinct operations in starting a slot that are all represented by their own method calls. Upon thinking about it further, it shouldn't be too surprising that we need the application to be aware of all of those lifecycle methods, though there is the potential for making that more clear in the class and making sure that sequence is initiated more sanely, but it does feel like it shouldn't be here.

Anyways, let me know how this looks and if you can re-attach your changes! Sorry for the goof, still getting used to the future to allow owners to push changes!

gunnarmorling commented:
> Do you still have those locally?

Yes; I've just pushed a new branch with them here: https://github.com/gunnarmorling/debezium/tree/DBZ-1238 (see the last two commits). Can you work this into the PR here?

gunnarmorling commented:
Hey @addisonj, thanks for the update. So could you do the following:

* Rebase it to latest master
* Apply the changes I did from the branch I linked above
* Add any other changes you'd like to do (if any)

Once that's done, I'll do one more review pass of it (need to bring back all the context to my memory as it has been a while and I'm getting old ;). Thanks a lot!

addisonj commented:
@gunnarmorling thanks for all that! like all those changes.

I rebased this and squashed your changes into this, so I think it should have everything. I didn't really see anything else as part of this commit, but obviously I think there are some follow on commits as we have touch on previously.

Let me know if there is anything else!

gunnarmorling commented:
Hey @Naros, I'm not sure whether you're aware of this pending change to the PG connector, it is about giving more capabilities to snapshotting. I have two questions for you based on it:

* can you evaluate whether there's lots of merge conflicts with your pending work for pgoutput? I'd expect that it's not too much, as your work should mostly *add* new classes, but I might be missing something
* can you build this PR branch and run it against the different wal2json versions deployed on RDS (IIRC, you had an instance set-up)? This changes the way the replication slot is set up, and as we had issues with that in the past, I'd like to avoid any potential regressions due to the way wal2json handles unsupported options (they officially "deprecated" some plug-in options, in reality passing those will cause a hard failure, though).

Thanks a lot!

gunnarmorling commented:
@addisonj, just to double-check: the proposed change wouldn't alter the current default behavior of the connector, correct?

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/snapshot/SnapshotterWrapper.java:
```diff
@@ -0,0 +1,36 @@
+public class SnapshotterWrapper {
+
+    private final Snapshotter snapshotter;
+    private final OffsetState offsetState;
```
This one is unused.

addisonj replied:
good catch, yeah

gunnarmorling replied:
Removed it.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -85,64 +92,48 @@ private PostgresReplicationConnection(Configuration config,
-                        "You cannot have multiple slots with the same name active for the same database",
-                        slotName, postgresPluginName, database());
-                throw new IllegalStateException();
+                this.createReplicationSlot();
```
Why is this one needed here? Is this _only_ for the case where it hasn't been created before by the `RecordsSnapshotProducer`?

addisonj replied:
yes, and I think also when temporary slots are used

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -161,32 +152,71 @@ else if (slotInfo.active()) {
     public ReplicationStream startStreaming() throws SQLException, InterruptedException {
+        // this ensures we have the right type of connection, creates the slot if needed,
+        // and sets the defaultStartingPos, so we must set it here
+        initConnection();
```
Seems superfluous as it's done again at line 218.

addisonj replied:
this is needed because `defaultStartingPos` is set by calling initConnection. The comment there should highlight it better. We then also have it do in the in the `startStreaming(Long offset)` because both of these methods are called in the app.

gunnarmorling replied:
@addisonj, it seems the same effect can be achieved by simply calling `return startStreaming(null);` then? The `offset` will then be set to `defaultStartingPos`.

gunnarmorling commented:
@Naros, did you see my message above? Would be great if you could check this change against RDS. Thanks!

Naros commented:
> can you evaluate whether there's lots of merge conflicts with your pending work for pgoutput? I'd expect that it's not too much, as your work should mostly add new classes, but I might be missing something

There may be a few conflicts, most notably around `PostgresReplicationConnection` but nothing that will be a problem to resolve.  A vast majority of my changes are new classes or additional methods being added to existing classes which should merge nicely.

> can you build this PR branch and run it against the different wal2json versions deployed on RDS (IIRC, you had an instance set-up)?

I'll give this a shot today and report back with my findings.

gunnarmorling commented:
Sweet; thanks, @Naros!

Naros commented:
**Postgres 9.6.3 R1**

* `PostgresConnectorIT#shouldReceiveChangesForChangePKColumnDefinition`
I don't believe this test failure is related specifically to this PR as it also fails on master.  I briefly looked at it and I couldn't really diagnose why it works with a local docker image of postgres 9.6 but fails when connecting to RDS.  Will need more time on this one to pin-point the underlying problem.

   The specific issue is the test expects key schema to return a null value for field `pk3` but instead it returns `<Field{name=pk3, index=1, schema=Schema{INT}}>`.

* `RecordsStreamProducerIT#testPassingStreamParams`
This test fails with `option "add-tables"="s1.should_stream" is unknown`.
Possible incompatibility?

* `RecordsStreamProducerIT#testPassingStreamMultipleParams`
This test fails with `option "filter-tables"="s2.should_not_stream" is unknown`.
Possible incompatibility?

* `RecordsStreamProducerIT#shouldReceiveChangesForTypeConstraints`
This likely could be a timing problem with RDS and the test suite as I can't reproduce this locally but the failure happens at https://github.com/debezium/debezium/blob/master/debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java#L598 where we expect the precision to be `6` but is still `5`.

**Postgres 10.6 R1**

* `RecordsSnapshotProducerIT#shouldGenerateSnapshotAndContinueStreaming`
This test fails on RDS stating that a logical replication slot named `debezium` already exists and that we cannot have multiple defined with the same name.  This appears to happen during the transition between Snapshot and Stream modes, inside `RecordsStreamProducer#start`.  This is not a problem locally with master against my local postgres 10.9 docker image.

* `ReplicationnConnectionIT#shouldReceiveAndDecodeIndividualChanges`
This test fails on RDS with error that table `table_without_pk` cannot be updated because it does not have a replica identity and publishes updates.  Strangely this runs against master and my local 10.9 docker instance.

Naros commented:
So while testing with RDS, there were a few changes I had to make in order to make this work smoothly.  The one that I think is pertinent to this PR is in `PostgresConnectorIT.java`

```
@Before
public void before() {
  TestHelper.create().dropReplicationSlot("debezium");
  initializeConnectorTestFramework();
}
```
I noticed there was problems with the replication slot not being dropped correctly between the tests.  For the short-term I explored the above change to the test and it seemed to have resolved the issues.

gunnarmorling commented:
Hey @Naros, thanks a lot for the analysis! Something to digest indeed. Of those issues you describe above, do any of them also apply to the current master? I.e. are any specifically introduced by this PR here?

gunnarmorling commented:
Hey @addisonj, howdy? Did you have a chance to look into my comments/questions above? Thanks!

addisonj commented:
@gunnarmorling yeah sorry missed those with all the conversation here. I addressed them all inline, but I won't have time to make changes until next week or so, but I think there is enough information there if you want to push ahead on the last details of this!

Thanks for all the feedback :)

Naros commented:
@gunnarmorling Some of them do, I believe i indicated in the wall of text for each point whether it was a problem for master.  If I didn't explicitly state whether it was or was not, then it wasn't.

gunnarmorling commented:
@Naros, hum, hum. So I'm struggling a bit to to grasp what's the take-away from all this; essentially, the question is: is there any regression introduced by this PR or not? This one has been dragging on for quite some time, and I'd love to have it off the slate eventually.

gunnarmorling commented:
Pushed one commit with some clean-up.

Naros commented:
Hi @gunnarmorling, sorry for the confusion as my prior comments were a bit misleading in retrospect.  

I took another look at this PR and from a PostgreSQL 9.6 perspective it LGTM.  The only concern I saw was when testing using the `postgres-10` profile, I noticed we had one test class that had two failures:

1. `RecordsSnapshotProducerIT#shouldGenerateSnapshotsForPartitionedTables`
2. `RecordsSnapshotProducerIT#shouldGenerateSnapshotAndContinueStreaming`

The first failure was a known problem on master so by rebasing this PR with master, that error should automatically be fixed as Jiri and I uncovered this when I began working on pgoutput.  The second failure is most puzzling because even running that test against a clean PG10 environment, it fails with the following:

```
2019-07-08 17:28:13,556 INFO   Postgres|test_server|records-stream-producer  Obtained valid replication slot ReplicationSlot [active=true, latestFlushedLsn=896796512, catalogXmin=1296]   [io.debezium.connector.postgresql.connection.PostgresConnection]
2019-07-08 17:28:13,557 ERROR  Postgres|test_server|records-snapshot-producer  unexpected exception   [io.debezium.connector.postgresql.RecordsSnapshotProducer]
org.apache.kafka.connect.errors.ConnectException: java.lang.IllegalStateException: A logical replication slot named 'debezium' for plugin 'wal2json' and database 'postgres' is already active on the server.You cannot have multiple slots with the same name active for the same database.
        at io.debezium.connector.postgresql.RecordsStreamProducer.start(RecordsStreamProducer.java:136)
        at io.debezium.connector.postgresql.RecordsSnapshotProducer.lambda$startStreaming$3(RecordsSnapshotProducer.java:158)
        at java.util.Optional.ifPresent(Optional.java:159)
        at io.debezium.connector.postgresql.RecordsSnapshotProducer.startStreaming(RecordsSnapshotProducer.java:148)
        at io.debezium.connector.postgresql.RecordsSnapshotProducer.lambda$start$1(RecordsSnapshotProducer.java:109)
        at java.util.concurrent.CompletableFuture.uniRun(CompletableFuture.java:705)
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:687)
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: A logical replication slot named 'debezium' for plugin 'wal2json' and database 'postgres' is already active on the server.You cannot have multiple slots with the same name active for the same database.
        at io.debezium.connector.postgresql.connection.PostgresReplicationConnection.ensureSlotNotActive(PostgresReplicationConnection.java:121)
        at io.debezium.connector.postgresql.connection.PostgresReplicationConnection.initReplicationSlot(PostgresReplicationConnection.java:136)
        at io.debezium.connector.postgresql.connection.PostgresReplicationConnection.initConnection(PostgresReplicationConnection.java:234)
        at io.debezium.connector.postgresql.connection.PostgresReplicationConnection.startStreaming(PostgresReplicationConnection.java:218)
        at io.debezium.connector.postgresql.RecordsStreamProducer.start(RecordsStreamProducer.java:118)
        ... 10 more
```

From what I can tell, the above is the only regression this PR introduces and is only applicable when running the test suite with the `postgres-10` profile.

gunnarmorling commented:
Hey @addisonj, so we did some more testing with this PR on Amazon RDS, and this revealed one issue (which isn't specific to RDS); specifically, when creating a *temporary* replication slot in the snapshotter (as it's the case during testing), it will be active right away. So when calling `ensureSlotNotActive()` during starting the streamer, it will raise an exception. This is the case for PG 10 onwards.

So I'm wondering: is the check in `ensureSlotNotActive()` perhaps a bit over-ambitious, and we could simply remove it? I'd expect that PG itself will complain when connecting to an already active slot a second time. WDYT?

gunnarmorling commented:
> The only concern I saw was when testing using the postgres-10 profile
>
> RecordsSnapshotProducerIT#shouldGenerateSnapshotAndContinueStreaming

Thanks, @Naros. That one is a problem indeed (see above). Let's wait to hear back from @addisonj.

gunnarmorling commented:
Pushed one more commit to remove `ensureSlotNotActive()`, as dicussed above.

gunnarmorling commented:
Pushed another commit for simplifying the handling around `startStreaming()`.

gunnarmorling commented:
@addisonj So I've pushed a few more commits, most notably to remove the check for the activeness of the replication slot (see the discussion above for the details). I think with this one it's good to go. 

Could you do the following:

* double-check on the commits I've added
* rebase the PR to latest master (there's one small conflict)
* send a PR against the doc repo, updating the [PG connector docs](https://github.com/debezium/debezium.github.io/blob/develop/docs/connectors/postgresql.asciidoc) so it mentions the `snapshot.mode` of `custom` together with the `snapshot.custom.class` option (in fact, we had missed that before). Ideally, that'd discuss how to use all this to take advantage of lock-free, exported snapshots

Thanks a lot again! Very happy to see this finally done, the support for exported snapshots is super-cool. It's my personal goal to iterate though such changes more quickly going forward, but it's not always easy given the complexity of the matter and many other things going on. So esp. thanks for your patience and hanging in there through all the reviewing rounds!

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -161,32 +133,65 @@ else if (slotInfo.active()) {
     @Override
     public ReplicationStream startStreaming() throws SQLException, InterruptedException {
-        return startStreaming(defaultStartingPos);
+        return startStreaming(null);
```
:+1: for this simplification

addisonj commented:
Okay, I looked at all those changes :+1:  from me on it, rebased as well, and I opened up the docs commit here: https://github.com/debezium/debezium.github.io/pull/327

gunnarmorling commented:
Thanks a lot, @addisonj! Very nice to see this done :)

gunnarmorling commented:
@addisonj, btw. are you using this for exported snapshots already? We were thinking to add a snapshot mode "exported" based on the new SPI out of the box (see [DBZ-1035](https://issues.jboss.org/browse/DBZ-1035)). Eventually, this even might become the default, if there are no downsides to using it.

addisonj commented:
Not yet! I haven't had time to experiment with it yet, but plan to do so in a few weeks when I get back to this.
